<a href="https://colab.research.google.com/github/Ishansingh438/EDA-Hotel-Booking-Analysis-project/blob/main/Ishan_Singh_Hotel_Booking_Analysis_Capstone_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## <b> Have you ever wondered when the best time of year to book a hotel room is? Or the optimal length of stay in order to get the best daily rate? What if you wanted to predict whether or not a hotel was likely to receive a disproportionately high number of special requests? This hotel booking dataset can help you explore those questions!

## <b>This data set contains booking information for a city hotel and a resort hotel, and includes information such as when the booking was made, length of stay, the number of adults, children, and/or babies, and the number of available parking spaces, among other things. All personally identifying information has been removed from the data. </b>

## <b> Explore and analyze the data to discover important factors that govern the bookings. </b>

Lets start...

## Hotel Booking Analysis:
This project dataset consists of the booking details of a City hotel and a Resort hotel from year 2015 to 2017.The main objectives of this project is to do exploratry data analysis from hotel point of view so that the management could make and ammend their policies accordingly.

This project analysis and visualize the dataset in five segments:

 1.Examining the guest distribution
 
 2.Exploring the bookings

 3.Inspecting the hotel traffic

 4.Examining special requests and amenities

 5.Studying the average daily rates

 On the basis of analysis of the above segments,required conclusions will be made.

# **Data Preparation and Cleaning**
1. First of all, I Downloaded the "hotel_bookings.csv" data file, and then I uplaoded the data from my computer to the colab notebook in order to be able to perform data analysis and get insight from this data.
2. I began by importing all the necessary python packages and libraries : I impored Pandas with the alias pd, imported Numpy with the alias np, imported matplotlib with the alias plt for the data visualization and imported Seaborn library with the alias sns for making statistical graphics.
3. I inserted the csv data into Padas dataframe and take some quick insight over the data.

In [3]:
#import pandas package
import pandas as pd
import numpy as np

#Matplotlib for additional customization
import matplotlib.pyplot as plt

# Seaborn for plotting and styling
import seaborn as sns


In [5]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# **Lets us read our data first**



In [6]:
# loading csv File
file_path = '/content/drive/MyDrive/Colab Notebooks/EDA Capstone Project/Copy of Hotel Bookings.csv'
hotel_df=pd.read_csv(file_path)
hotel_df

hotel  is_canceled  lead_time  arrival_date_year  \
0       Resort Hotel            0        342               2015   
1       Resort Hotel            0        737               2015   
2       Resort Hotel            0          7               2015   
3       Resort Hotel            0         13               2015   
4       Resort Hotel            0         14               2015   
...              ...          ...        ...                ...   
119385    City Hotel            0         23               2017   
119386    City Hotel            0        102               2017   
119387    City Hotel            0         34               2017   
119388    City Hotel            0        109               2017   
119389    City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
0                    July                        27   
1                    July                        27   
2                    July                        27   
3                    July                        27   
4                    July                        27   
...                   ...                       ...   
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  ...  deposit_type  agent company  \
0                          0       2  ...    No Deposit    NaN     NaN   
1                          0       2  ...    No Deposit    NaN     NaN   
2                          1       1  ...    No Deposit    NaN     NaN   
3                          1       1  ...    No Deposit  304.0     NaN   
4                          2       2  ...    No Deposit  240.0     NaN   
...                      ...     ...  ...           ...    ...     ...   
119385                     5       2  ...    No Deposit  394.0     NaN   
119386                     5       3  ...    No Deposit    9.0     NaN   
119387                     5       2  ...    No Deposit    9.0     NaN   
119388                     5       2  ...    No Deposit   89.0     NaN   
119389                     7       2  ...    No Deposit    9.0     NaN   

       days_in_waiting_list customer_type     adr  \
0                         0     Transient    0.00   
1                         0     Transient    0.00   
2                         0     Transient   75.00   
3                         0     Transient   75.00   
4                         0     Transient   98.00   
...                     ...           ...     ...   
119385                    0     Transient   96.14   
119386                    0     Transient  225.43   
119387                    0     Transient  157.71   
119388                    0     Transient  104.40   
119389                    0     Transient  151.20   

        required_car_parking_spaces  total_of_special_requests  \
0                                 0                          0   
1                                 0                          0   
2                                 0                          0   
3             

By the first glance at the dataset, we can see that the table containse 32 columns of booking information over 119390 hotels.
By using the .shape() method we can get the same kind of information without the table:

In [7]:
hotel_df.shape

(119390, 32)

In [8]:
hotel_df.columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'adults', 'children', 'babies', 'meal',
       'country', 'market_segment', 'distribution_channel',
       'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type', 'agent',
       'company', 'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests',
       'reservation_status', 'reservation_status_date'],
      dtype='object')

Now lets take a high-level look over the data using the .info() method:

In [9]:
hotel_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

It appears that each of the 32 columns contains values of a specific data type (int64, float64 or object),however the Non-Null Count reveals that some of the features (that are less than 119390) contains missing information.

Even, by presenting the first 8 rows of the dataset, we can notice right away the apperance of the miising values at the "agent" and the "company" colomns.

In [10]:
hotel_df.head(8)

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   
5  Resort Hotel            0         14               2015               July   
6  Resort Hotel            0          0               2015               July   
7  Resort Hotel            0          9               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   
5                        27                          1   
6                        27                          1   
7                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  deposit_type  \
0                        0                     0       2  ...    No Deposit   
1                        0                     0       2  ...    No Deposit   
2                        0                     1       1  ...    No Deposit   
3                        0                     1       1  ...    No Deposit   
4                        0                     2       2  ...    No Deposit   
5                        0                     2       2  ...    No Deposit   
6                        0                     2       2  ...    No Deposit   
7                        0                     2       2  ...    No Deposit   

   agent company days_in_waiting_list customer_type    adr  \
0    NaN     NaN                    0     Transient    0.0   
1    NaN     NaN                    0     Transient    0.0   
2    NaN     NaN                    0     Transient   75.0   
3  304.0     NaN                    0     Transient   75.0   
4  240.0     NaN                    0     Transient   98.0   
5  240.0     NaN                    0     Transient   98.0   
6    NaN     NaN                    0     Transient  107.0   
7  303.0     NaN                    0     Transient  103.0   

   required_car_parking_spaces  total_of_special_requests  reservation_status  \
0                            0                          0           Check-Out   
1                            0                          0           Check-Out   
2                            0                          0           Check-Out   
3                            0                          0           Check-Out   
4                            0                          1           Check-Out   
5                            0                          1           Check-Out   
6                            0                          0           Check-Out   
7                            0                          1           Check-Out   

  reservation_status_date  
0              2015-07-01  
1              2015-07-01  
2              2015-07-02  
3              2015-07-02  
4              2015-07-03  
5              2015-07-03  
6              2015-07-03  
7              2015-07-03  

[8 rows x 32 columns]

let us find out more missing values in table using .isnull() method 

In [11]:
# find the columns that have nan values 
nan_values = hotel_df.isnull().sum().sort_values(ascending=False)
# print only features that have values greater than 0 (filter out zero value )
nan_values[nan_values>0]

company     112593
agent        16340
country        488
children         4
dtype: int64

# **Replacing Missing Values**
The above analysis shows that the dataset contains 4 colums with a missing value (NAN) as follow:

### ♦company   =  112593
### ♦agent     =   16340
### ♦country    =    488
### ♦children   =      4

Lets clean the data by swaping the NAN/mising values with a real numbers:

### 1. Replacing missing values for the "country" feature
Since the countries contains categorical numbers (represented in the ISO 3155–3:2013 format, we should find the relative number of hotels in each country. Than we should replace the total missing value of countries with other countries according to their overall weight.

In [12]:
hotel_df['country'] = hotel_df['country'].fillna(0)
hotel_df['country'].head(5)

0    PRT
1    PRT
2    GBR
3    GBR
4    GBR
Name: country, dtype: object

### 2. Replacing missing values for the "agent" and "company" features
The agent and company features represent the ID number of the travel agency and the ID number of the company/entity that made the booking. In this particular fields we can asume that in most cases nan value means that the booking or reservarsion has been done without any agent or company, so we will simply replace the missing values with 0.

By Looking at the hotel_df.info() we can see that the data type of both segment and company features is float64.
We will preserve this type of dataset by replacing the null values id of the agent or company in the copied dataset with 0.0.

In [13]:
hotel_df['company'] = hotel_df['company'].fillna(0.0)
hotel_df['company'].head()

0    0.0
1    0.0
2    0.0
3    0.0
4    0.0
Name: company, dtype: float64

In [14]:
hotel_df['agent'] = hotel_df['agent'].fillna(0.0)
hotel_df['agent'].head()

0      0.0
1      0.0
2      0.0
3    304.0
4    240.0
Name: agent, dtype: float64

# 3. Replacing missing values for the "children" feature
To replace the missing values in the "children" column, I will swap the nan values of children with the closest integer number of the average number of children, we can get this number using the .describe() method that can give us more statistic details on all the table's features or just by using the .mean() method to get this particular number.

In [15]:
hotel_df.describe()

is_canceled      lead_time  arrival_date_year  \
count  119390.000000  119390.000000      119390.000000   
mean        0.370416     104.011416        2016.156554   
std         0.482918     106.863097           0.707476   
min         0.000000       0.000000        2015.000000   
25%         0.000000      18.000000        2016.000000   
50%         0.000000      69.000000        2016.000000   
75%         1.000000     160.000000        2017.000000   
max         1.000000     737.000000        2017.000000   

       arrival_date_week_number  arrival_date_day_of_month  \
count             119390.000000              119390.000000   
mean                  27.165173                  15.798241   
std                   13.605138                   8.780829   
min                    1.000000                   1.000000   
25%                   16.000000                   8.000000   
50%                   28.000000                  16.000000   
75%                   38.000000                  23.000000   
max                   53.000000                  31.000000   

       stays_in_weekend_nights  stays_in_week_nights         adults  \
count            119390.000000         119390.000000  119390.000000   
mean                  0.927599              2.500302       1.856403   
std                   0.998613              1.908286       0.579261   
min                   0.000000              0.000000       0.000000   
25%                   0.000000              1.000000       2.000000   
50%                   1.000000              2.000000       2.000000   
75%                   2.000000              3.000000       2.000000   
max                  19.000000             50.000000      55.000000   

            children         babies  is_repeated_guest  \
count  119386.000000  119390.000000      119390.000000   
mean        0.103890       0.007949           0.031912   
std         0.398561       0.097436           0.175767   
min         0.000000       0.000000           0.000000   
25%         0.000000       0.000000           0.000000   
50%         0.000000       0.000000           0.000000   
75%         0.000000       0.000000           0.000000   
max        10.000000      10.000000           1.000000   

       previous_cancellations  previous_bookings_not_canceled  \
count           119390.000000                   119390.000000   
mean                 0.087118                        0.137097   
std                  0.844336                        1.497437   
min                  0.000000                        0.000000   
25%                  0.000000                        0.000000   
50%                  0.000000                        0.000000   
75%                  0.000000                        0.000000   
max                 26.000000                       72.000000   

       booking_changes          agent        company  days_in_waiting_list  \
count    119390.000000  119390.000000  119390.000000         119390.000000   
mean          0.221124      74.828319      10.775157              2.321149   
std           0.652306     107.141953      53.943884             17.594721   
min           0.000000       0.000000       0.000000              0.000000   
25%           0.000000       7.000000       0.000000              0.000000   
50%           0.000000       9.000000       0.000000              0.000000   
75%           0.000000     152.000000       0.000000              0.000000   
max          21.000000     535.000000     543.000000            391.000000   

                 adr  required_car_parking_spaces  total_of_special_requests  
count  119390.000000                119390.000000              119390.000000  
mean      101.831122                     0.062518                   0.571363  
std        50.535790                     0.245291                   0.792798  
min        -6.380000                     0.000000                   0.000000  
25%        69.290000                     0.000000                   0.000000  
5

In [16]:
hotel_df['children'].mean()

0.10388990333874994

From this analysis we can see that the avarage value of number of children is around 0.1
Looking at the "hotels_df.info()" above, we can see that the "children" type is float64, so in order to keep this format we will replace each of the four NAN missing children values with the number: 0.0 using the .fillna() method.

We will replace values in the copy of the original table:

In [17]:
hotel_df = pd.read_csv(file_path)
hotel_df_2 = hotel_df.copy()
nan_values_2 = hotel_df_2.isnull().sum().sort_values(ascending=False)


In [18]:
# Print only features that have values greater than 0 (filter out zeros value )
nan_values_2[nan_values_2>0]

company     112593
agent        16340
country        488
children         4
dtype: int64

In [19]:
hotel_df_2['children'].fillna(0.0 , inplace = True)

In [20]:
# Print only features that have values greater than 0 (filter out zeros value )
nan_values_2 = hotel_df_2.isnull().sum().sort_values(ascending=False)

nan_values_2[nan_values_2>0]

company    112593
agent       16340
country       488
dtype: int64

## 4. Dealing with Unexpected Values
Several rows in the dataset contains values that does not make any sense like having no adults, children, and babies. We will replace the amount of adults within this rows with the rounded avarage adults values.

First of all let's find this rows:

In [21]:
no_guest = hotel_df_2.adults + hotel_df_2.children + hotel_df_2.babies
no_guest[no_guest ==0]

2224      0.0
2409      0.0
3181      0.0
3684      0.0
3708      0.0
         ... 
115029    0.0
115091    0.0
116251    0.0
116534    0.0
117087    0.0
Length: 180, dtype: float64

In [22]:
# Replacing the zero adults with the avarage adults number and round it
adults_mean_value = hotel_df_2['adults'].mean()
print(round(adults_mean_value))

2


In [23]:
# lets look at the adult columns after the replacment
hotel_df_2.loc[:, 'adults'] = round(adults_mean_value)
round(hotel_df_2)
hotel_df_2['adults'].head()

0    2
1    2
2    2
3    2
4    2
Name: adults, dtype: int64

## Dealing with total  of "special requests" by hotel type



In [51]:
# Total of special requests is greater than zero 
hotel_df_2 = hotel_df_2[hotel_df_2['total_of_special_requests'] > 0]
print(hotel_df_2.shape)

(49072, 32)


In [87]:
# See how many special requests by hotel type
print('city hotel')
total_of_special_requests_city_cnt = hotel_df_2[hotel_df_2.hotel == 'City Hotel'].total_of_special_requests.value_counts().sum()
print (total_of_special_requests_city_cnt) 
print('resort hotel')
total_of_special_requests_resort_cnt = hotel_df_2[hotel_df_2.hotel == 'Resort Hotel'].total_of_special_requests.value_counts().sum()
print (total_of_special_requests_resort_cnt)

city hotel
31373
resort hotel
17699


# Is canceled


In [66]:
# See how many canceled the reservation by hotel type
print('City Hotel:')

# See how many special requests reservation canceled by City hotel
is_canceled_city = hotel_df_2[hotel_df_2.hotel == 'City Hotel'].is_canceled.sum()
print('Canceled:',is_canceled_city)

print('Resort Hotel')

# See how many special requests reservation canceled by Resort hotel
is_canceled_resort = hotel_df_2[hotel_df_2.hotel == 'Resort Hotel'].is_canceled.sum()
print('Canceled:',is_canceled_resort)


City Hotel:
Canceled: 6762
Resort Hotel
Canceled: 3906
